### PR TITLE
Add metadata tables to Raptor

### PR DESCRIPTION
**Pull request by nileema**

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+        Long tableId = tableIds.next();
+        Long columnId = metadataDao.getTemporalColumnId(tableId);
+        String minColumn = (columnId == null) ? "null" : minColumn(columnId);
+        String maxColumn = (columnId == null) ? "null" : maxColumn(columnId);
+
+        try {
+            connection = dbi.open().getConnection();
+            statement = preparedStatementBuilder.createPreparedStatement(connection, minColumn, maxColumn, shardIndexTable(tableId));
+            System.out.println("Executing query: " + this.statement);    // TODO <============ remove this line
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Remove this

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                return value;
+            }
+            if (value instanceof String) {
+                return Slices.wrappedBuffer(((String) value).getBytes(UTF_8));
+            }
+            if (value instanceof byte[]) {
+                return Slices.wrappedBuffer((byte[]) value);
+            }
+        }
+        throw new IllegalArgumentException(String.format("Type %s incompatible with %s", type, value));
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Static import

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    private static Object normalizeValue(Type type, @Nullable Object value)
+    {
+        Class<?> javaType = type.getJavaType();
+        if (javaType.isPrimitive()) {
+            checkArgument(value == null || Primitives.wrap(javaType).isInstance(value), "Type %s incompatible with %s", type, value);
+            return value;
+        }
+        else if (javaType == Slice.class) {
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Redundant else

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+public class ValueBuffer
+{
+    private final int columnIndex;
+    private final Type type;
+    private final Object value;
+
+    private ValueBuffer(int columnIndex, Type type, @Nullable Object value)
+    {
+        this.columnIndex = columnIndex;
+        this.type = type;
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Check for null

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        checkState(!isNull());
+        return value;
+    }
+
+    public int getColumnIndex()
+    {
+        return columnIndex;
+    }
+
+    @SuppressWarnings("ConstantConditions")
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Why are these here?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        ValueBuffer other = (ValueBuffer) obj;
+        return Objects.equal(this.type, other.type) && Objects.equal(this.value, other.value);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Formatting

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        return type;
+    }
+
+    public boolean isNull()
+    {
+        return value == null;
+    }
+
+    @Override
+    public int hashCode()
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Put `hashCode` after `equals`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        return Objects.hashCode(type, value);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ValueBuffer.java]: Use JDK not Guava

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        if (javaType == boolean.class) {
+            return Types.BOOLEAN;
+        }
+        if (javaType == Slice.class) {
+            return Types.VARBINARY;
+        }
+        throw new IllegalArgumentException("Unknown type: " + type);
+    }
+
+    private static void enableStreamingResults(Statement statement)
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: Move this to a utility rather than copying

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Iterables.limit;
+import static java.lang.String.format;
+import static java.sql.ResultSet.CONCUR_READ_ONLY;
+import static java.sql.ResultSet.TYPE_FORWARD_ONLY;
+import static java.util.Objects.requireNonNull;
+
+public class PreparedStatementBuilder
+{
+    private final List<ValueBuffer> bindValues = new ArrayList<>(256);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: Why 256?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.uuidColumnIndexes = ImmutableList.copyOf(requireNonNull(uuidColumnIndexes, "uuidColumnIndexes is null"));
+
+        this.sqlTemplate = sqlTemplate + getWhereClause(tupleDomain);
+    }
+
+    PreparedStatement createPreparedStatement(Connection connection, String... args)
+            throws SQLException
+    {
+        String sql = format(sqlTemplate, args);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: I think it would be cleaner to pass in the SQL

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        }
+
+        return "(" + Joiner.on(" OR ").join(disjuncts) + ")";
+    }
+
+    private Object getBindValue(int columnIndex, Object value)
+    {
+        if (uuidColumnIndexes.contains(columnIndex)) {
+            UUID uuid = UUID.fromString(new String(((Slice) value).getBytes()));
+            return UuidUtil.uuidToBytes(uuid);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: Static import and inline

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+            disjuncts.add(columnName + " IS NULL");
+        }
+
+        return "(" + Joiner.on(" OR ").join(disjuncts) + ")";
+    }
+
+    private Object getBindValue(int columnIndex, Object value)
+    {
+        if (uuidColumnIndexes.contains(columnIndex)) {
+            UUID uuid = UUID.fromString(new String(((Slice) value).getBytes()));
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: `((Slice) value).toStringUtf8()`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 14]
```diff
@@ -983,7 +983,6 @@ public PhysicalOperation visitTableScan(TableScanNode node, LocalExecutionPlanCo
             for (Symbol symbol : node.getOutputSymbols()) {
                 columns.add(node.getAssignments().get(symbol));
             }
-
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: Unrelated change

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+        this.longs = new long[types.size()];
+        this.doubles = new double[types.size()];
+        this.strings = new String[types.size()];
+        this.nulls = new boolean[types.size()];
+    }
+
+    int extractValues(ResultSet resultSet, List<Integer> uuidColumns)
+            throws SQLException
+    {
+        verify(resultSet != null, "resultSet is null");
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java]: Should be `checkArgument`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+                nulls[i] = resultSet.wasNull();
+                if (!nulls[i]) {
+                    completedBytes += SIZE_OF_DOUBLE;
+                }
+            }
+            else if (javaType == Slice.class) {
+                if (uuidColumns.contains(i)) {
+                    byte[] bytes = resultSet.getBytes(i + 1);
+                    nulls[i] = resultSet.wasNull();
+                    strings[i] = nulls[i] ? null : uuidFromBytes(bytes).toString().toLowerCase();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java]: `ENGLISH`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                    nulls[i] = resultSet.wasNull();
+                    strings[i] = nulls[i] ? null : value;
+                }
+
+                if (!nulls[i]) {
+                    completedBytes += strings[i].length();
+                }
+            }
+            else {
+                throw new IllegalStateException("Unknown Java type: " + javaType);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java]: throw `PrestoException(INTERNAL_ERROR)`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+        requireNonNull(dbi, "dbi is null");
+        requireNonNull(tupleDomain, "tupleDomain is null");
+
+        this.dbi = dbi;
+        this.metadataDao = dbi.onDemand(MetadataDao.class);
+
+        this.tableIds = getTableIds(metadataDao, tupleDomain);
+
+        List<String> columnNames = createQualifiedColumnName();
+        this.preparedStatementBuilder = new PreparedStatementBuilder(constructSqlTemplate(columnNames), columnNames, TYPES, ImmutableList.of(COLUMNS.indexOf(SHARD_UUID_COLUMN)), tupleDomain);
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Wrap arguments

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+        StringBuilder sql = new StringBuilder();
+        sql.append("SELECT\n");
+
+        for (int i = 0; i < columnNames.size() - 2; i++) {
+            sql.append(columnNames.get(i) + ",\n");
+        }
+        sql.append("%s AS min_timestamp,\n");
+        sql.append("%s AS max_timestamp\n");
+
+        sql.append("FROM %s x JOIN shards\nON (x.shard_id = shards.shard_id)\n");
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Put `ON` on next line

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+
+        return sql.toString();
+    }
+
+    private static List<String> createQualifiedColumnName()
+    {
+        ImmutableList.Builder<String> columnNames = ImmutableList.builder();
+        String shards = "shards";
+        String tables = "tables";
+        columnNames.add(tables + "." + COLUMNS.get(0).getName());
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Just inline these

Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: This can be a single chained `return builder()...build()`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+                resultSet = getNextResultSet();
+                if (resultSet == null) {
+                    close();
+                    return false;
+                }
+            }
+            completedBytes += resultSetValues.extractValues(resultSet, ImmutableList.of(COLUMNS.indexOf(SHARD_UUID_COLUMN)));
+            return true;
+        }
+        catch (Exception e) {
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: This should catch the proper exceptions

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+
+    private boolean closed;
+    private Connection connection;
+    private PreparedStatement statement;
+    private ResultSet resultSet;
+    private long completedBytes;
+
+    public ShardMetadataRecordCursor(IDBI dbi)
+    {
+        this(dbi, TupleDomain.<Integer>all());
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Is `<Integer>` necessary?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java (hunk 17 above)
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ResultSetValues.java]: throw `VerifyException`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+        }
+        else if (type.getJavaType() == Slice.class) {
+            preparedStatement.setString(parameterIndex, new String(valueBuffer.getSlice().getBytes()));
+        }
+        else {
+            throw new IllegalStateException("Unknown Java type: " + type.getJavaType());
+        }
+    }
+
+    private static int typeToSqlType(Type type)
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: add one for VarChar

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java (hunk 10 above)
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: arbitrary

**Comment by electrum** [comment]
Did you mean to close this?

**Comment by electrum** [comment]
Move the `SystemTable` changes for `TupleDomain` into a separate commit.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+        sql.append("SELECT\n");
+
+        for (int i = 0; i < columnNames.size() - 2; i++) {
+            sql.append(columnNames.get(i) + ",\n");
+        }
+        sql.append("%s AS min_timestamp,\n");
+        sql.append("%s AS max_timestamp\n");
+
+        sql.append("FROM %s x JOIN shards\nON (x.shard_id = shards.shard_id)\n");
+        sql.append("JOIN tables on (shards.table_id = tables.table_id)");
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Add newline at end of string

**Comment by nileema** [comment]
@electrum - Addressed comments and rebased to trunk

**Comment by electrum** [comment]
Looks good
